### ISSUE TITLE
NullReferenceException in EnsureRequestIdHandlerAttribute

### ISSUE THREAD

**Issue by suwatch** [projectkudu/kudu#2547]
Encountered while investigating API Plane Health.

```json
{"Message":"An error has occurred.","ExceptionMessage":"Object reference not set to an instance of an object","ExceptionType":"System.NullReferenceException","StackTrace":"
  at Kudu.Services.EnsureRequestIdHandlerAttribute.OnActionExecuted (System.Web.Http.Filters.HttpActionExecutedContext actionExecutedContext) [0x00047] in <e144cc92a33944beb179bd9981d2fa89>:0 \n
  at System.Web.Http.Filters.ActionFilterAttribute.OnActionExecutedAsync (System.Web.Http.Filters.HttpActionExecutedContext actionExecutedContext, System.Threading.CancellationToken cancellationToken) [0x00000] in <f99f496cb0d249c1a945c1fcabce1695>:0 \n
--- End of stack trace from previous location where exception was thrown ---\n
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <4dc8ec68b0964e099af86e50301f5f3c>:0 \n
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <4dc8ec68b0964e099af86e50301f5f3c>:0 \n
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <4dc8ec68b0964e099af86e50301f5f3c>:0 \n
```

**Comment by nickwalkmsft** [comment]
@davidebbo shared this same stack trace with me a few days ago, he was able to generate it on a Linux site in a specific scenario; I'll let him share more details. I haven't looked into it in detail yet. Do you happen to know if the offending sites were Linux?

**Comment by davidebbo** [comment]
In my case, I was calling `/api/functions` on Linux site and hit this, but no idea what caused it. @suwatch, I assume you saw this on Windows? Let's get to the bottom of it as it blocked me from making progress on my Functions/Linux experiments.

**Comment by suwatch** [comment]
Just about to add that it is only happening on Linux workers.  Try Kusto below for details.   They are from /siteextensions api.

```
AntaresAdminGeoEvents
| where PreciseTimeStamp >= datetime(2017-09-01)
| where PreciseTimeStamp <= datetime(2017-09-06)
| where EventId == 50004
| where Details contains "EnsureRequestIdHandlerAttribute" and Details contains "NullRef"
| summarize count() by SiteName, Address
```

**Comment by nickwalkmsft** [comment]
Thanks. I will raise priority on this.

**Comment by davidebbo** [comment]
In my case it was functions and not siteextension, but I'm guessing the issue is agnostic of the specific API.

As an aside, I'm not sure why anyone is calling siteextensions api on Linux!

**Comment by suwatch** [comment]
> As an aside, I'm not sure why anyone is calling siteextensions api on Linux!

It is done thru ARM surface with browser User-Agent; so highly likely, from Portal.   If this api is not applicable for Linux, we have a couple options to address this (such as fail with 4xx or empty array from GeoMaster).   Regardless, it would good to get to the bottom of the null ref.

**Comment by davidebbo** [comment]
Yes, let's keep in mind that I ran into this with Functions API, and that one is relevant on Linux (or will be).

**Comment by nickwalkmsft** [comment]
Update on this: the null ref is coming from the access of [HttpContext.Current](https://github.com/projectkudu/kudu/blob/master/Kudu.Services/EnsureRequestIdHandlerAttribute.cs#L20-L26), and it's null because something about [searchResource.Search](https://github.com/projectkudu/kudu/blob/bb683db7bb748aaa49d1e062c59f1258e0827166/Kudu.Core/SiteExtensions/FeedExtensions.cs#L61) is causing loss of context on Mono. I'm not sure exactly what to do about this; running it synchronously with .Result works, but is hacky.

I'm also not sure where else this may be occurring.

**Comment by suwatch** [comment]
@nickwalkmsft  why not just check for null (add comment) and return for now to get us moving?

**Comment by nickwalkmsft** [comment]
Yep, that's the plan. Did something similar in https://github.com/projectkudu/kudu/pull/2333.

**Comment by davidebbo** [comment]
Yep, that makes sense. At this point, I just want to avoid this crash so we can get to the next step.

**Comment by nickwalkmsft** [comment]
https://github.com/projectkudu/kudu/pull/2556

**Comment by jvano** [comment]
Hi

If the problem persists and is related to running it on Azure App Service, please open a support incident in Azure:
https://learn.microsoft.com/en-us/azure/azure-portal/supportability/how-to-create-azure-support-request

This way we can better track and assist you on this case

Thanks,

Joaquin Vano
Azure App Service
